akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -218,3 +225,191 @@ private[fusing] object StreamOfStreams {
     }
   }
 }
+
+/**
+ * INTERNAL API
+ */
+object PrefixAndTail {
+
+  sealed trait MaterializationState
+  case object NotMaterialized extends MaterializationState
+  case object AlreadyMaterialized extends MaterializationState
+  case object TimedOut extends MaterializationState
+
+  case object NormalCompletion extends MaterializationState
+  case class FailureCompletion(ex: Throwable) extends MaterializationState
+
+  trait TailInterface[T] {
+    def pushSubstream(elem: T): Unit
+    def completeSubstream(): Unit
+    def failSubstream(ex: Throwable)
+  }
+
+  final class TailSource[T](
+    timeout: FiniteDuration,
+    register: TailInterface[T] ⇒ Unit,
+    pullParent: Unit ⇒ Unit,
+    cancelParent: Unit ⇒ Unit) extends GraphStage[SourceShape[T]] {
```

Comment:
Then the line `tailSource = new TailSource[T](timeout, onSubstreamRegister.invoke, onSubstreamPull.invoke, onSubstreamFinish.invoke)` will not compile though. Strange it is.